common/flatpak-run.c

```diff
@@ -3322,8 +3322,46 @@ flatpak_run_app (const char     *app_ref,
 
   if (app_deploy != NULL)
     {
+      g_autofree const char **previous_ids = NULL;
+      gsize len = 0;
+
       app_files = flatpak_deploy_get_files (app_deploy);
 
+      g_autoptr (GFile) current_data_dir = flatpak_get_data_dir (app_ref_parts[1]);
+      previous_ids = flatpak_deploy_data_get_previous_ids (app_deploy_data, &len);
+      while (len > 0)
+        {
+          g_autoptr (GFile) previous_data_dir = NULL;
+          g_autoptr (GError) local_error = NULL;
+          g_autofree char *previous_data_path = NULL;
+          g_autofree char *current_data_path = NULL;
+
+          if (g_file_query_exists (current_data_dir, cancellable))
+            break;
+
+          previous_data_dir = flatpak_get_data_dir (previous_ids[--len]);
+          if (!g_file_query_exists (previous_data_dir, cancellable))
+            continue;
+
+          if (!flatpak_file_rename (previous_data_dir,
+                                    current_data_dir,
+                                    cancellable, &local_error))
```

Comment:
I wonder if this should be a hard error, or give some sort of warning? I can imagine if you did something weird like putting the steam app data on a different mount where this failed, and now you can't ever run the app anymore.